project_cost_report/report.py

```diff
@@ -437,3 +439,96 @@ def get_foldable_categories(self, project_id):
         res = super().get_foldable_categories(project_id)
         res['outsourcing'] = [False]
         return res
+
+
+class ProjectCostReportWithTotalCost(models.TransientModel):
+
+    _inherit = 'project.cost.report'
+
+    def _get_rendering_variables(self, project, report_context):
+        """Add the total of costs to the rendering variables."""
+        res = super()._get_rendering_variables(project, report_context)
+        res.update({
+            'total_cost': float_round(
+                res['product_total'] + res['time_total'] + res['outsourcing_total'], 2
+            )
+        })
+        return res
+
+
+def purchase_line_is_waiting_invoice(line: models.Model) -> bool:
+    """Determine whether a purchase line is waiting an invoice.
+
+    :param line: a purchase.order.line singleton
+    """
+    precision = line.env['decimal.precision'].precision_get('Product Unit of Measure')
+    less_units_invoiced_than_purchased = (
+        float_compare(line.qty_invoiced, line.product_qty, precision_digits=precision) == -1
+    )
+    return less_units_invoiced_than_purchased
+
+
+def get_purchase_line_waiting_qty(line: models.Model) -> float:
+    """Get the quantity of units waiting invoices from a purchase line.
+
+    :param line: a purchase.order.line singleton
+    """
+    precision = line.env['decimal.precision'].precision_get('Product Unit of Measure')
+    return float_round(line.product_qty - line.qty_invoiced, precision_digits=precision)
+
+
+def get_waiting_for_invoice_total(order: models.Model, project: models.Model):
+    """Get the total amount waiting for invoices for a purchase order.
+
+    :param order: the purchase order record.
+    :param project: the project for which to render the report.
+    :return: the amount waiting for invoices
+    """
+    lines_waiting_invoices = order.order_line.filtered(
+        lambda l: purchase_line_is_waiting_invoice(l) and
+        l.account_analytic_id == project.analytic_account_id
+    )
+
+    return float_round(
+        sum(l.price_unit * get_purchase_line_waiting_qty(l)
+            for l in lines_waiting_invoices),
+        2
+    )
+
+
+class ProjectCostReportWithWaitingInvoices(models.TransientModel):
+
+    _inherit = 'project.cost.report'
+
+    def _get_waiting_purchase_order_lines(self, project):
+        """Get the purchase order lines with unreceived invoices.
+
+        :param project: the project.project record
+        :rtype: purchase.order.line
+        """
+        domain = [
+            ('account_analytic_id', '=', project.analytic_account_id.id),
+            ('order_id.state', 'in', ('purchase', 'done')),
+        ]
+        lines = self.env['purchase.order.line'].search(domain)
+        return lines.filtered(lambda l: purchase_line_is_waiting_invoice(l))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Lambda may not be necessary](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=3730559)